### ISSUE TITLE
Publishing FRR Debian based appliance

### ISSUE THREAD

**Issue by elico** [GNS3/gns3-registry#697]
Since the FRR VM appliance is stuck for 8 month on the 8.1 version I want to publish a Debian based FRR appliance which is supported by the official FRR team.
I am looking for some directions to guidance on the right way to create and publish an appliance in the registry.

It's pretty simple for me to create a VM but I am not sure on how to clean it up and prepare the system and the networking configuration for publishing.
Any help would be great.

Thanks,

**Comment by giangi** [comment]
HI @elico, not a part of the GNS3 team, but I was looking into this myself. Have you read https://www.uni-koeln.de/~pbogusze/posts/Building_64bit_alpine_linux_GNS3_FRRouting_appliance.html ? I found it relatively clear and convincing and was going to take a shot at creating a new version for the Alpine appliance using the [8.4.0 packages in edge](https://pkgs.alpinelinux.org/packages?name=frr&branch=edge&repo=&arch=&maintainer=).

I was particularly interested in Alpine because of the small footprint, but having to pull packages from edge was a little of a turn-off for me. Using the official FRR stable Debian packages seems reasonable.

Are you confident that you could do it with the guidance of the blog post I suggested, or would you like to see if I can pull it off? I admit that while I am relatively confident that I could build an image that works, I have zero experience with appliance/image submission, so any help there would still be needed.

**Comment by giangi** [comment]
Hi @elico. The [README](https://github.com/GNS3/gns3-registry#policy-for-new-appliances) states

> We welcome pull requests for new appliances.
> 
> Regarding images / disks referenced in new appliance files, we accept links that point to well-known vendor websites or other trusted source websites.
> 
> For new contribution, we accept links depending on multiple criteria: the kind of appliance, who submits it and especially if scripts to build the images(s) are provided (packer scripts are recommended for Qemu appliances). Then we will build on our side, verify and upload the image(s) on the GNS3 Sourceforge account.
> 
> For Docker appliances, please provide a Dockerfile.

I am also interested, and have some technical knowledge, but no experience on the right way to build one. I was looking at copying from what already present in the [packer directory](https://github.com/GNS3/gns3-registry/tree/master/packer), and maybe integrating some suggestions from https://www.uni-koeln.de/~pbogusze/posts/Building_64bit_alpine_linux_GNS3_FRRouting_appliance.html.

This should help with both updating the current Alpine-based appliance and create a new Debian one. I have no experience with packer so I cannot guarantee results. Do you feel that you can accomplish this easily with your skillset, or would you like me to try? Obviously any correction and further guidance is appreciated.

**Comment by elico** [comment]
@giangi the Frr is pretty simple to do since it really only needs to be updated to the latest 3.16 and maybe to also offer a edge version which seems pretty simple.
I will need to verify that I am able to do so and it will take me a week or more since I don't have a lot of spare time.

About the debian part, I will try to look for a packer example and then will try to verify how I to do so.
There is another possibility with debian and it's to create a cloud-init iso that will configure the existing VM and will install FRR.
I can implement the cloud-init solution pretty fast but it would not be a fully "offline" appliance.

Lets try to be in touch here in the next 2 weeks to make sure we can do this.

Thanks,

**Comment by giangi** [comment]
Hi. I could not resist the urge to scratch the itch, and as a weekend project I tried to whip up something in [giangi/frr-gns3](https://github.com/giangi/frr-gns3). It may not be that good, but I will have less time once the working week starts, so I wanted to shared it before then.

* I tried to play around with packer as a first experience. I have some experience with Debian's preseeding, so I started with a standard image.
* I also wanted to try using Debian's cloud images because their smaller size was appealing, but I am much less confident in my ability to use cloud-init, and to integrate it properly with packer. I kind of cobbled together an alternative solution using libguestfs as an intermediate step.
* I probably absolutely butchered the whole `.gns3a` concept, but for a quick test I thought it was acceptable

I was able to import both appliances into GNS3 and boot them, but I did not go into any real test. Maybe they are a starting point, and if they are good enough a more stable contribution could be made here using some ideas from them. Probably the first thing is to confirm that they work and fit the title. I will try to replace some of the 8.1.0 Alpine appliances from the marketplace in one of my project with these experimental Debian ones to test them further. Any feedback is appreciated.

**Comment by elico** [comment]
@giangi I will try to look at it next week.
Looks pretty good.
I myself was looking for Debian preseeding instructions but didn't managed yet to even scratch it.
I do like RHEL based systems and feel stable with them.

**Comment by giangi** [comment]
Thanks @b-ehlers . Your work helped me muster the courage to go through the cloud-init documentation and showed me how to integrate it in packer. I updated my experiments at [giangi/frr-gns3](https://github.com/giangi/frr-gns3).
I did not have time to make any significant test apart from basic configuration, but the images seem in a good overall shape.

May I ask how one would proceed with the contribution? I expect that doing a pull request with a packer setup for the cloud image variant like you did for the Debian Cloud base image would be a first step. Did I understand correctly from the README that once such contribution has been made, I could rely on the GNS3 team to create the official `.gns3a` template file, build images, uploading them to the GNS3 sourceforge account, and do whatever is necessary for the marketplace integration?
